## Week 5 Homework 

In this homework we'll put what we learned about Spark in practice.

For this homework we will be using the FHVHV 2021-06 data found here. [FHVHV Data](https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhvhv/fhvhv_tripdata_2021-06.csv.gz )


### Question 1: 

**Install Spark and PySpark** 

- Install Spark
- Run PySpark
- Create a local spark session
- Execute spark.version.

What's the output?
- 3.3.2
- 2.1.4
- 1.2.3
- 5.4
</br></br>

R:/ 3.3.2

In [6]:
!spark-shell

23/03/08 06:40:57 WARN Utils: Your hostname, DESKTOP-LRRD21B resolves to a loopback address: 127.0.1.1; using 172.17.195.122 instead (on interface eth0)
23/03/08 06:40:57 WARN Utils: Set SPARK_LOCAL_IP if you need to bind to another address
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
23/03/08 06:41:13 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
23/03/08 06:41:14 WARN Utils: Service 'SparkUI' could not bind on port 4040. Attempting port 4041.
Spark context Web UI available at http://172.17.195.122:4041
Spark context available as 'sc' (master = local[*], app id = local-1678275674103).
Spark session available as 'spark'.
Welcome to
      ____              __
     / __/__  ___ _____/ /__
    _\ \/ _ \/ _ `/ __/  '_/
   /___/ .__/\_,_/_/ /_/\_\   version 3.3.2
      /_/
         
Using Scala version 2.12.15 (OpenJDK 64-Bit Ser

In [18]:
import pyspark
from pyspark.sql import SparkSession
from pyspark.sql import types
from pyspark.sql import functions as F

In [4]:
spark = SparkSession.builder \
    .master("local[*]") \
    .appName('test') \
    .getOrCreate()

your 131072x1 screen size is bogus. expect trouble


23/03/08 06:55:07 WARN Utils: Your hostname, DESKTOP-LRRD21B resolves to a loopback address: 127.0.1.1; using 172.17.195.122 instead (on interface eth0)
23/03/08 06:55:07 WARN Utils: Set SPARK_LOCAL_IP if you need to bind to another address


Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).


23/03/08 06:55:17 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable


### Question 2: 

**HVFHW June 2021**

Read it with Spark using the same schema as we did in the lessons.</br> 
We will use this dataset for all the remaining questions.</br>
Repartition it to 12 partitions and save it to parquet.</br>
What is the average size of the Parquet (ending with .parquet extension) Files that were created (in MB)? Select the answer which most closely matches.</br>


- 2MB
- 24MB
- 100MB
- 250MB

R:/ 24 MB
</br></br>

In [2]:
schema = types.StructType([
    types.StructField('hvfhs_license_num', types.StringType(), True),
    types.StructField('dispatching_base_num', types.StringType(), True),
    types.StructField('pickup_datetime', types.TimestampType(), True),
    types.StructField('dropoff_datetime', types.TimestampType(), True),
    types.StructField('PULocationID', types.IntegerType(), True),
    types.StructField('DOLocationID', types.IntegerType(), True),
    types.StructField('SR_Flag', types.StringType(), True)
])

In [13]:
df = spark.read \
    .option("header", "true").option("inferSchema", "true") \
    .csv('fhvhv_tripdata_2021-06.csv')

df = df.repartition(12)

df.repartition(12).write.mode("overwrite").parquet('data/pq/fhvhv/2021/06/')

23/03/08 07:04:46 WARN MemoryManager: Total allocation exceeds 95.00% (908,486,235 bytes) of heap memory
Scaling row group sizes to 96.70% for 7 writers
23/03/08 07:04:46 WARN MemoryManager: Total allocation exceeds 95.00% (908,486,235 bytes) of heap memory
Scaling row group sizes to 84.61% for 8 writers
23/03/08 07:04:46 WARN MemoryManager: Total allocation exceeds 95.00% (908,486,235 bytes) of heap memory
Scaling row group sizes to 75.21% for 9 writers
23/03/08 07:04:46 WARN MemoryManager: Total allocation exceeds 95.00% (908,486,235 bytes) of heap memory
Scaling row group sizes to 67.69% for 10 writers
23/03/08 07:04:46 WARN MemoryManager: Total allocation exceeds 95.00% (908,486,235 bytes) of heap memory
Scaling row group sizes to 61.53% for 11 writers
23/03/08 07:04:46 WARN MemoryManager: Total allocation exceeds 95.00% (908,486,235 bytes) of heap memory
Scaling row group sizes to 56.41% for 12 writers


23/03/08 07:04:49 WARN MemoryManager: Total allocation exceeds 95.00% (908,486,235 bytes) of heap memory
Scaling row group sizes to 61.53% for 11 writers
23/03/08 07:04:49 WARN MemoryManager: Total allocation exceeds 95.00% (908,486,235 bytes) of heap memory
Scaling row group sizes to 67.69% for 10 writers
23/03/08 07:04:49 WARN MemoryManager: Total allocation exceeds 95.00% (908,486,235 bytes) of heap memory
Scaling row group sizes to 75.21% for 9 writers
23/03/08 07:04:49 WARN MemoryManager: Total allocation exceeds 95.00% (908,486,235 bytes) of heap memory
Scaling row group sizes to 84.61% for 8 writers
23/03/08 07:04:49 WARN MemoryManager: Total allocation exceeds 95.00% (908,486,235 bytes) of heap memory
Scaling row group sizes to 96.70% for 7 writers


### Question 3: 

**Count records**  

How many taxi trips were there on June 15?</br></br>
Consider only trips that started on June 15.</br>

- 308,164
- 12,856
- 452,470
- 50,982

R: 452,470
</br></br>

In [16]:
df2 = spark.read \
    .option("header", "true").option("inferSchema", "true") \
    .parquet('data/pq/fhvhv/2021/06/')

In [17]:
df2.show(5)

+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+
|dispatching_base_num|    pickup_datetime|   dropoff_datetime|PULocationID|DOLocationID|SR_Flag|Affiliated_base_number|
+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+
|              B02510|2021-06-01 16:24:08|2021-06-01 17:25:06|         129|          69|      N|                  null|
|              B02510|2021-06-01 12:48:33|2021-06-01 13:03:00|         229|         100|      N|                  null|
|              B02879|2021-06-01 07:20:27|2021-06-01 08:10:56|          68|         265|      N|                B02879|
|              B02889|2021-06-02 10:00:42|2021-06-02 10:09:50|         164|          79|      N|                B02889|
|              B02872|2021-06-03 11:38:01|2021-06-03 11:59:39|         162|          79|      N|                B02872|
+--------------------+------------------

In [20]:
df2 \
    .withColumn('pickup_date', F.to_date(df.pickup_datetime)) \
    .filter("pickup_date = '2021-06-15'") \
    .count()

452470

### Question 4: 

**Longest trip for each day**  

Now calculate the duration for each trip.</br>
How long was the longest trip in Hours?</br>

- 66.87 Hours
- 243.44 Hours
- 7.68 Hours
- 3.32 Hours
</br></br>

In [27]:
df2.printSchema()
df2.createOrReplaceTempView ('fhvhv_2021_06')

root
 |-- dispatching_base_num: string (nullable = true)
 |-- pickup_datetime: timestamp (nullable = true)
 |-- dropoff_datetime: timestamp (nullable = true)
 |-- PULocationID: integer (nullable = true)
 |-- DOLocationID: integer (nullable = true)
 |-- SR_Flag: string (nullable = true)
 |-- Affiliated_base_number: string (nullable = true)



In [30]:
spark.sql("""
    SELECT
        to_date(pickup_datetime) AS pickup_date,
        MAX((CAST(dropoff_datetime AS LONG) - CAST(pickup_datetime AS LONG)) / 60) AS duration
    FROM 
        fhvhv_2021_06
    GROUP BY
        pickup_date
    ORDER BY
        duration DESC
    LIMIT 10;
""").show()

+-----------+------------------+
|pickup_date|          duration|
+-----------+------------------+
| 2021-06-25| 4012.733333333333|
| 2021-06-22|1532.9833333333333|
| 2021-06-27|           1198.85|
| 2021-06-26|1091.8333333333333|
| 2021-06-23| 988.0166666666667|
| 2021-06-24| 834.5833333333334|
| 2021-06-04|             700.2|
| 2021-06-20| 659.0666666666667|
| 2021-06-01|            616.05|
| 2021-06-28| 597.9833333333333|
+-----------+------------------+



### Question 5: 

**User Interface**

 Spark’s User Interface which shows application's dashboard runs on which local port?</br>

- 80
- 443
- 4040
- 8080

R: 4040
</br></br>


### Question 6: 

**Most frequent pickup location zone**

Load the zone lookup data into a temp view in Spark</br>
[Zone Data](https://github.com/DataTalksClub/nyc-tlc-data/releases/download/misc/taxi_zone_lookup.csv)</br>

Using the zone lookup data and the fhvhv June 2021 data, what is the name of the most frequent pickup location zone?</br>

- East Chelsea
- Astoria
- Union Sq
- Crown Heights North

R: Crown Heights North
</br></br>


In [32]:
zones_df = spark.read \
    .option("header", "true").option("inferSchema", "true") \
    .csv('taxi_zone_lookup.csv')

In [35]:
df2.show(5)

+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+
|dispatching_base_num|    pickup_datetime|   dropoff_datetime|PULocationID|DOLocationID|SR_Flag|Affiliated_base_number|
+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+
|              B02510|2021-06-01 16:24:08|2021-06-01 17:25:06|         129|          69|      N|                  null|
|              B02510|2021-06-01 12:48:33|2021-06-01 13:03:00|         229|         100|      N|                  null|
|              B02879|2021-06-01 07:20:27|2021-06-01 08:10:56|          68|         265|      N|                B02879|
|              B02889|2021-06-02 10:00:42|2021-06-02 10:09:50|         164|          79|      N|                B02889|
|              B02872|2021-06-03 11:38:01|2021-06-03 11:59:39|         162|          79|      N|                B02872|
+--------------------+------------------

In [34]:
zones_df.show(5)

+----------+-------------+--------------------+------------+
|LocationID|      Borough|                Zone|service_zone|
+----------+-------------+--------------------+------------+
|         1|          EWR|      Newark Airport|         EWR|
|         2|       Queens|         Jamaica Bay|   Boro Zone|
|         3|        Bronx|Allerton/Pelham G...|   Boro Zone|
|         4|    Manhattan|       Alphabet City| Yellow Zone|
|         5|Staten Island|       Arden Heights|   Boro Zone|
+----------+-------------+--------------------+------------+
only showing top 5 rows



In [38]:
joined_df = df2.join(
    zones_df,
    df2.PULocationID ==  zones_df.LocationID,
    "inner"
)

joined_df.show(5)

+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+----------+---------+--------------------+------------+
|dispatching_base_num|    pickup_datetime|   dropoff_datetime|PULocationID|DOLocationID|SR_Flag|Affiliated_base_number|LocationID|  Borough|                Zone|service_zone|
+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+----------+---------+--------------------+------------+
|              B02510|2021-06-01 16:24:08|2021-06-01 17:25:06|         129|          69|      N|                  null|       129|   Queens|     Jackson Heights|   Boro Zone|
|              B02510|2021-06-01 12:48:33|2021-06-01 13:03:00|         229|         100|      N|                  null|       229|Manhattan|Sutton Place/Turt...| Yellow Zone|
|              B02879|2021-06-01 07:20:27|2021-06-01 08:10:56|          68|         265|      N|                B02879|      

In [46]:
joined_df.groupBy('Zone').count().orderBy('count', ascending=False).show()

+--------------------+------+
|                Zone| count|
+--------------------+------+
| Crown Heights North|231279|
|        East Village|221244|
|         JFK Airport|188867|
|      Bushwick South|187929|
|       East New York|186780|
|TriBeCa/Civic Center|164344|
|   LaGuardia Airport|161596|
|            Union Sq|158937|
|        West Village|154698|
|             Astoria|152493|
|     Lower East Side|151020|
|        East Chelsea|147673|
|Central Harlem North|146402|
|Williamsburg (Nor...|143683|
|          Park Slope|143594|
|  Stuyvesant Heights|141427|
|        Clinton East|139611|
|West Chelsea/Huds...|139431|
|             Bedford|138428|
|         Murray Hill|137879|
+--------------------+------+
only showing top 20 rows



## Submitting the solutions

* Form for submitting: https://forms.gle/EcSvDs6vp64gcGuD8
* You can submit your homework multiple times. In this case, only the last submission will be used. 

Deadline: 06 March (Monday), 22:00 CET


## Solution

We will publish the solution here